CMake improvements

haydenroche5 opened the pull request:
- Begin adding options to enable/disable different features.
- Increase minimum CMake version to 3.2.
- Support installation of the built files.
- Add checks for necessary include files, functions etc.
- Generate options.h and config.h.

dgarske commented on CMakeLists.txt:
```diff
@@ -226,32 +1494,104 @@ if(BUILD_TESTS)
 endif()
 
-# TODO: Add install() for library, headers and test applications
+# Build Unit Tests
```
The `--disable-examples` option disables the unit test as well.

haydenroche5 replied:
Got it, fixed.

dgarske reviewed (changes requested):
Hey Hayden,

Looks like you'll need to add `cmake/config.in` and `cmake/functions.cmake` to Makefile.am.

Please document the Cmake arguments somewhere... and add example for use in INSTALL. Also let's keep the cmake variable cache simple like `-DARC4=1`.

Thanks

haydenroche5 commented:
> Hey Hayden,
> 
> Looks like you'll need to add `cmake/config.in` and `cmake/functions.cmake` to Makefile.am.
> 
> Please document the Cmake arguments somewhere... and add example for use in INSTALL. Also let's keep the cmake variable cache simple like `-DARC4=1`.
> 
> Thanks

Seeing as there's no --help option-type thing for CMake, I've pointed users to cmake-gui  and CMakeCache.txt for documentation on what the various configuration options are and what they do in INSTALL. Definitely open to other ideas for how to make these options more visible, if you have any ideas.

dgarske reviewed (changes requested):
@haydenroche5  Wow, awesome work on this. 

A couple of notes and items to consider before its gets merged:
1) If there were previously a cmake . done in the root then CMakeFiles and CMakeCache.txt need removed otherwise it will report "In-source builds are not allowed.".

2) Can we better document an example for enable/disable options? Such as disabling DH `cmake .. -DWOLFSSL_DH=no` and note the build file `wolfssl/options.h` is still generated and can be included in end application.

3) Let's consider putting the final applications in the same place autoconf does. As long as a make clean will remove those we should be fine.

Thanks! David

haydenroche5 commented:
> @haydenroche5 Wow, awesome work on this.
> 
> A couple of notes and items to consider before its gets merged:
> 
> 1. If there were previously a cmake . done in the root then CMakeFiles and CMakeCache.txt need removed otherwise it will report "In-source builds are not allowed.".
> 2. Can we better document an example for enable/disable options? Such as disabling DH `cmake .. -DWOLFSSL_DH=no` and note the build file `wolfssl/options.h` is still generated and can be included in end application.
> 3. Let's consider putting the final applications in the same place autoconf does. As long as a make clean will remove those we should be fine.
> 
> Thanks! David

Thanks for the feedback, David.

1. Annoyingly, CMake creates that directory and file after CMakeLists.txt execution ends, so I can't clean them up in the CMakeLists.txt. I've modified the error message to let the user know that they need to delete these manually if they try an in-source build. See https://medium.com/@qwertie/cmake-fights-against-disabling-in-source-builds-ab1d71c1d26f.
2. I've added additional documentation to INSTALL to make this stuff more clear. Let me know if you think it should go elsewhere, though.
3. Done!

dgarske reviewed (approved):
Perfect! Excellent CMake additions. One Jenkins is working we can get this merged.
